kraft-liveness-readiness.md

```diff
@@ -0,0 +1,112 @@
+for a Kafka cluster that is using KRaft rather than ZooKeeper. This includes a KRaft 
+cluster with combined nodes and one that has separate controller and broker nodes.
+
+## Current situation
+
+These are the current probes used in Strimzi when ZooKeeper is present:
+
+|Component| Liveness | Readiness |
+|---|---|---|
+| ZooKeeper | "imok" from 127.0.0.1:12181 | "imok" from 127.0.0.1:12181 |
+| ZK-based broker | if (ready) { listening on 9091 } else { have ZK session } | if (ready) { return true } |
+
+The "ready" check in both cases is based on the existence of the file `../kafka-ready` which is 
+created by the KafkaAgent when the broker state metric has value 3 (RUNNING).
+
+## Background
+
+This section contains some background information about the ports and metrics we could use for checks.
+
+### Ports
+In KRaft mode there are two different ports that the node could be listening on.
+If it has a controller process it should be listening on the port defined as the `controller.listener.names` which in Strimzi is currently hard-coded to 9090.
+If it has a broker process it should be listening on the replication port which is hard-coded to 9091.
+If the node is running as a combined mode, it should be listening on both ports
+
+### BrokerState
+The expected [BrokerState](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java) transitions are:
+
+NOT_RUNNING (0) -> STARTING (1) -> RECOVERY (2) -> RUNNING (3) -> PENDING_CONTROLLED_SHUTDOWN (6) -> SHUTTING_DOWN (7)
+
+The other possible value is 127 which represents UNKNOWN.
+
+### Current state metric and quorum
+
+Some useful things to be aware of with the current-state metric and how the controller quorum affects the broker nodes:
+
+* The [current-state](https://kafka.apache.org/documentation/#kraft_quorum_monitoring) metric indicates whether a controller is currently a leader or follower.
+If this metric shows one of these two states, this means the leader election has happened and a controller quorum has been successfully formed.
+* The current-state metric is not accessible to the existing KafkaAgent because it is not a Yammer metric.
+* The broker nodes will not move to a RUNNING state until the quorum leader election has happened.
+
+## Proposal
+
+There are three different types of liveness/readiness probe that we need to consider, first for a combined node, 
+then for a broker and then for a controller.
+
+### Combined mode
+
+The proposed probes are:
+
+* Liveness: node is listening on 9090 and 9091
+* Readiness: node is listening on 9090 and 9091
+
+### Controller only mode
+
+The proposed probes are:
+
+* Liveness: node is listening on 9090
+* Readiness: node is listening on 9090
+
+### Broker only mode
+
+The proposed probes are:
+
+* Liveness: the BrokerState metric < 3 || listening on 9091
+* Readiness: the BrokerState metric >=3 && != 127
```

Comment:
Why is this not part of the combined node readiness as well? For example in an similar form to how it works today with ZooKeeper when the Readiness probe essentially changes after reaching some point?

Reply:
I've explained in the rejected alternatives section why the BrokerState metric check isn't included in the combined node readiness check. Does that answer your question sufficiently or is there something still unclear there?